### PR TITLE
ASGIRequest compatible middleware

### PR DESCRIPTION
**Pull request by an7oine**
Super call handles MiddlewareMixin._async_check() too.

### CODE REVIEW

**Review by LeMinaw — COMMENTED** [review]
Looking at [the source for MiddlewareMixin init method](https://github.com/django/django/blob/main/django/utils/deprecation.py#L88), it seems the reason no supercall is made ATM is the parent class does not support `None` as a `get_response` argument (which is set by the subclass constructor as a default value).

I'm not familiar enough with django-hosts to be sure of the implications though.

**Comment by an7oine** [comment]
Crucial part is `self._async_check()`. This could be done here if supercall is a problem.

**Comment by LeMinaw** [comment]
I'll investigate a bit on this in the next few days, there is still something very unclear about this `get_response` parameter to me.

Perhaps in the meantime, it would be great to provide some testcase(s?) involving `ASGIRequest` / async capability?

**Review by LeMinaw — APPROVED** [review]
Thanks, looks good to me!

**Comment by an7oine** [comment]
Are we getting close to a merge? Ability to install from Pypi in ASGI production would be nice.

**Comment by ddabble** [comment]
@LeMinaw
> Looking at [the source for MiddlewareMixin init method](https://github.com/django/django/blob/main/django/utils/deprecation.py#L88), it seems the reason no supercall is made ATM is the parent class does not support `None` as a `get_response` argument (which is set by the subclass constructor as a default value).

It seems like back when the `get_response` argument was added (in [this commit](https://github.com/jazzband/django-hosts/commit/7f63567c7b6f42cda9d86c194195ff7fea146e67#diff-4f2b737c1aebe597cd447dd6e4adf9ac69bd8fc31714a69d1c21bba818b5ced9R24)), it pretty much copied the code in [the parent class' `__init__()` method](https://github.com/django/django/blob/69de988f9237fe978c5fca363ca2d2622f8e2ccd/django/utils/deprecation.py#L115-L117), and so maybe not calling the `__init__()` method was an oversight instead of a deliberate choice to allow a `None` value..? 🤔 Regardless, as Django's code for `MiddlewareMixin` stands today, it has potential for producing an error if `get_response` is `None` (both in [`__call__()`](https://github.com/django/django/blob/4.0/django/utils/deprecation.py#L126) and [`__acall__()`](https://github.com/django/django/blob/4.0/django/utils/deprecation.py#L142)), which is not expected behavior of a child class. Also, I'm fairly sure (my knowledge on Django middleware is also limited 😅) that the `get_response` argument is only provided by Django itself when initializing the middleware chain - with no practical way of changing this behavior - and that it's not reasonable to assume that users of this library will somehow try to pass `None`. If this is *not* the case, however, we should probably just add a line to the changelog saying that the `get_response` argument doesn't accept a `None` value 🙂

File: django_hosts/middleware.py [hunk 1]
```diff
@@ -15,7 +15,7 @@ class HostsBaseMiddleware(MiddlewareMixin):
     toolbar_middleware = 'debug_toolbar.middleware.DebugToolbarMiddleware'
 
     def __init__(self, get_response=None):
```
Comment (ddabble) [inline django_hosts/middleware.py:17]: We can remove the default value now that Django's `__init__()` method requires a non-null value 🙂
```suggestion
    def __init__(self, get_response):
```

Comment (ddabble) [reply django_hosts/middleware.py:19]: I forgot that Django < 4.0 still supports this default argument value, so nvm 😅 I instead added a deprecation warning comment 🙂

**Comment by ddabble** [comment]
Current release 5.0 is incompatible with [channels](https://pypi.org/project/channels/) >= 3.0.0 (as stated in the changelog in cc0f448e5e00fb5042dcc7415603c0c3e1d368b7), as the middleware classes are not correctly marked as async-capable, which this PR fixes (by calling `MiddlewareMixin.__init__()`), so it would be nice to have this PR merged and version 5.1 released 🙂

Do you have any input on this, @timgraham?

File: tests/test_middleware.py [hunk 2]
```diff
@@ -1,3 +1,15 @@
+import asyncio
+from unittest import skipUnless
+
+# Not guaranteed / nonexistent for Django below 3.0/3.1.
```
Comment (timgraham) [inline tests/test_middleware.py:4]: The minimum supported version is now 3.2, so please simplify.

File: tests/test_middleware.py [hunk 3]
```diff
@@ -9,24 +21,34 @@
 from .base import HostsTestCase
 
 
+def mock_get_response(request):
```
Comment (timgraham) [inline tests/test_middleware.py:24]: In `django/tests/middleware/tests.py`, a similar method is called `get_response_empty` so I'd suggest that name.

**Review by timgraham — COMMENTED** [review]
Are the commits meant to be squashed? If not, please structure them appropriately.

**Comment by ddabble** [comment]
I squashed all 6 commits into 1 (203cf3bd5f1faa7b697e622f553c6e8a051152cf), as they all contained tightly connected changes 🙂

@timgraham Is this more like what you had in mind?

**Comment by ddabble** [comment]
There doesn't seem to be much activity from any other experienced contributors here, so would you have time to review the changes I made that you requested, @timgraham? 🙂

**Pull request merged**